torch/onnx/_internal/fx/fx_exporter.py

```diff
@@ -1,23 +1,297 @@
+    depending on the tracing strategy.
+    """
+
+    def adapt(self, model_outputs: Any) -> Sequence[Any]:
+        """Flatten the model outputs."""
+        flattened_outputs, _ = pytree.tree_flatten(model_outputs)
+        return flattened_outputs
+
+
+class FlattenOutputWithTreeSpecValidationStep:
+    """Same as ``FlattenOutputStep``, with additional `TreeSpec` validation.
+
+    This class stores the `SpecTree` output produced when `adapt` was called the first
+    time. It then validates the `SpecTree` output produced from later `adapt` calls.
+    """
+
+    _spec: Optional[pytree.TreeSpec] = None
+
+    def adapt(self, model_outputs: Any) -> Sequence[Any]:
+        """Flatten the model outputs and validate the `SpecTree` output.
+
+        Args:
+            model_outputs: The model outputs to flatten.
+
+        Returns:
+            flattened_outputs: The flattened model outputs.
+
+        Raises:
+            ValueError: If the `SpecTree` output produced from the current `model_outputs`
+                is not identical to the `SpecTree` output produced from the first
+                `model_outputs` that was passed to this method.
+        """
+        flattened_outputs, spec = pytree.tree_flatten(model_outputs)
+        if self._spec is None:
+            self._spec = spec
+        else:
+            _assert_identical_pytree_spec(
+                self._spec,
+                spec,
+                error_message="Model outputs incompatible with the format that was exported. ",
+            )
+        return flattened_outputs
+
+
+class FXGraphModuleExporter(exporter.Exporter, abc.ABC):
+    _input_adapter: exporter.InputAdapter
+    _output_adapter: exporter.OutputAdapter
+
     @property
     def decomposition_table(self) -> Mapping[torch._ops.OpOverload, Callable]:
         return function_dispatcher._ONNX_FRIENDLY_DECOMPOSITION_TABLE
 
+    def _apply_input_adapt_step(
+        self,
+        adapt_step_cls: Type[exporter.InputAdaptStep],
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+        step_init_args: Optional[Sequence[Any]] = None,
+    ) -> Tuple[Sequence[Any], Mapping[str, Any]]:
+        """Apply an input adapt step to the model args and kwargs.
+
+        An input adapt step object is initialized, applied and recorded as part of
+        ``self._input_adapter`.
+
+        Args:
+            adapt_step_cls: The input adapt step class.
+            model_args: The model args.
+            model_kwargs: The model kwargs.
+            step_init_args: The input adapt step initialization arguments.
+
+        Returns:
+            The adapted model args and kwargs.
+        """
+        step_init_args = step_init_args or ()
+        adapt_step = adapt_step_cls(*step_init_args)
+        self._input_adapter.append_step(adapt_step)
+        return adapt_step.adapt(model_args, model_kwargs)
+
+    def _apply_output_adapt_step(
```

Comment:
we could move all these adapters to a file of its own and declutter the exporter class files.

Reply:
Sure for follow ups